### DISCUSSION TITLE
Is it possible to have inline typeguards?

### DISCUSSION THREAD

**Discussion by rzvc** [vimeo/psalm#7948] [Q&A]
This is what I would like to do, without the need of defining a function that executes that particular condition:

```php
$arr = ['foo' => 'abcd', 'bar' => 123];

foreach ($arr as $k => $v)
{
    /** @psalm-assert-if-true 'abcd' $v */
    if ($k == 'foo')
    {
        /** @psalm-trace $v */
        echo $v;
    }
}
```
https://psalm.dev/r/c07ff98909

In this case, ideally I would like it to figure out that $v is type 'abcd', but lacking that, an inline type guard would be nice.

**Comment by orklah** [comment]
How would it be different than just using @var ?
https://psalm.dev/r/3f133055db

**Reply by rzvc** [reply]
It's the same thing but the intention is more obvious, because it would show where the type is being decided. This would also look better:
```php
$arr = ['foo' => 'abcd', 'bar' => 123];

foreach ($arr as $k => $v)
{
    /** @psalm-assert-if-false 123 $v */
    if ($k == 'foo')
        continue;

    // $v is now int.
}
```
vs the alternative:
```php
$arr = ['foo' => 'abcd', 'bar' => 123];

foreach ($arr as $k => $v)
{
    if ($k == 'foo')
        continue;

    /** @var 123 $v */
    // $v is now int. But when I read this, I understand it more as the
    // result of an imperative action rather than a question/answer.
}
```

**Comment by AndrolGenhald** [comment]
Use `assert`? https://psalm.dev/r/f1d55ee169
The only downside is it only supports type assertions you can do in PHP, so if you want to assert that something is an array shape it'll get quite verbose.
